smac/intensification/successive_halving.py

```diff
@@ -0,0 +1,464 @@
+        """
+        initialize Successive Halving parameters
+
+        Parameters
+        ----------
+        initial_budget : typing.Optional[float]
+            minimum budget allowed for 1 run of successive halving
+        max_budget : typing.Optional[float]
+            maximum budget allowed for 1 run of successive halving
+        eta : float
+            'halving' factor after each iteration in a successive halving run
+        num_initial_challengers : typing.Optional[int]
+            number of challengers to consider for the initial budget
+        """
+
+        if eta <= 1:
+            raise ValueError('eta must be greater than 1')
+        self.eta = eta
+
+        # BUDGETS
+
+        if max_budget is not None and initial_budget is not None \
+                and max_budget < initial_budget:
+            raise ValueError('Max budget has to be larger than min budget')
+
+        # - if only 1 instance was provided & quality objective, then use cutoff as budget
+        # - else, use instances as budget
+        if not self.run_obj_time and len(self.instances) <= 1:
+            # budget with cutoff
+            if initial_budget is None or \
+                    (max_budget is None and self.cutoff is None):
+                raise ValueError("Successive Halving with runtime-cutoff as budget (i.e., only 1 instance) "
+                                 "requires parameters initial_budget and max_budget/cutoff for intensification!")
+
+            if self.cutoff is not None and max_budget is not None:
+                self.logger.warning('Successive Halving with runtime-cutoff as budget: '
+                                    'Both max budget (%d) and runtime-cutoff (%d) were provided. '
+                                    'Max budget will be used.' % (max_budget, self.cutoff))
+
+            self.initial_budget = initial_budget
+            self.max_budget = max_budget if max_budget else self.cutoff
+            self.cutoff_as_budget = True
+
+        else:
+            # budget with instances
+            if self.run_obj_time and len(self.instances) <= 1:
+                self.logger.warning("Successive Halving has objective 'runtime' but only 1 instance-seed pair.")
+            self.initial_budget = 1 if initial_budget is None else int(initial_budget)
+            self.max_budget = len(self.instances) if max_budget is None else int(max_budget)
+            self.cutoff_as_budget = False
+
+            if self.max_budget > len(self.instances):
+                raise ValueError('Max budget cannot be greater than the number of instance-seed pairs')
+            if self.max_budget < len(self.instances):
+                self.logger.warning('Max budget (%d) does not include all instance-seed pairs (%d)' %
+                                    (self.max_budget, len(self.instances)))
+
+        self.logger.debug("Running Successive Halving with '%s' as budget" % self.cutoff_as_budget)
```

Comment:
Could you the initial budget and the max budget to the log message?

Reply:
Could you please also add the decision process on how to race to the docstring of this class? Otherwise, I believe that it will be hard for users to understand what's happening.